### PR TITLE
last 

### PR DESCRIPTION
**Pull request by sunjincheng121**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/package.scala [hunk 1]
```diff
@@ -15,15 +15,40 @@
...
+import org.apache.flink.types.Row
+import org.apache.flink.api.scala._
+import org.apache.flink.api.scala.DataSet
+import org.apache.flink.streaming.api.scala.DataStream
+import org.apache.flink.table.api.Table
+import org.apache.flink.table.api.scala.{ImplicitExpressionConversions, DataSetConversions, DataStreamConversions, TableConversions, BatchTableEnvironment => ScalaBatchTableEnv, StreamTableEnvironment => ScalaStreamTableEnv}
+
+import _root_.scala.language.implicitConversions
+
+package object expressions extends ImplicitExpressionConversions {
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/package.scala]: 这个文件好像多余了，有个`package object scala extends ImplicitExpressionConversions`了

File: a.a [hunk 2]
```diff
@@ -0,0 +1 @@
+dd
```
Comment (hequn8128) [inline a.a]: 这个文件删了？

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/package.scala (hunk 1 above)
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/package.scala]: 这是为了解决依赖 ExpressionUtils 会使用这个，ExpressionUtils在table-common将来，package object scala extends ImplicitExpressionConversions 在org.apache.flink.table.api

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Alias.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Alias expression which maintain alias information for an Expression.
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Alias.java]: Expression -> expression
maintain -> maintains

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/FunctionDefinition.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * The base class of function definition which can directly define built-in functions
+ * with a name attribute.
+ */
+public class FunctionDefinition {
+	private String name;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/FunctionDefinition.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * The distinct aggregate expression.
+ */
+public class DistinctAgg extends UnaryExpression {
+
+	private Expression child;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Call.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The function call expression.
+ */
+public class Call extends Expression {
+
+	private FunctionDefinition func;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Call.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Literal.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+import java.util.Optional;
+
+/**
+ * Literal expression.
+ */
+public class Literal extends LeafExpression {
+
+	private Object value;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Literal.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Null.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.flink.table.expressions;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+/**
+ * Null expression.
+ */
+public class Null extends LeafExpression {
+
+	private TypeInformation<?> type;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Null.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/ProctimeAttribute.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.apache.flink.table.expressions;
+
+/**
+ * Proctime attribute expression for indicating, accessing, and working in
+ * Flink's processing time.
+ */
+public class ProctimeAttribute extends UnaryExpression {
+	private Expression expr;
+
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/ProctimeAttribute.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/RowtimeAttribute.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * Rowtime attribute for indicating, accessing, and working in Flink's event time.
+ */
+public class RowtimeAttribute extends UnaryExpression {
+
+	private Expression expr;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/RowtimeAttribute.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/ScalarFunctionDefinition.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.flink.table.expressions;
+
+import org.apache.flink.table.functions.ScalarFunction;
+
+/**
+ * The definition for user-defined scalar function.
+ */
+public class ScalarFunctionDefinition extends FunctionDefinition {
+
+	private ScalarFunction func;
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/ScalarFunctionDefinition.java]: final

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TypeLiteral.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+/**
+ * TypeLiteral which can wraps the typeInformation.
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TypeLiteral.java]: wraps -> wrap

Comment (dianfu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TypeLiteral.java]: typeInformation -> TypeInformation

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/UDAGGExpression.scala [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.flink.table.plan.expressions
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils.{getAccumulatorTypeOfAggregateFunction, getResultTypeOfAggregateFunction}
+
+/**
+  * A class which creates a call to an aggregateFunction
+  */
+case class UDAGGExpression[T: TypeInformation, ACC: TypeInformation](
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/UDAGGExpression.scala]: 这个文件好像可以删了

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala [hunk 14]
```diff
@@ -47,114 +48,114 @@ trait ImplicitExpressionOperations {
     *
     * e.g. 12.toExpr % 'a
     *
-    * @return expression
+    * @return Expression
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala]: unnecessary change, revert?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala [hunk 15]
```diff
@@ -576,21 +578,16 @@ trait ImplicitExpressionOperations {
     *   .select('c, 'a, 'a.count over 'w, 'a.sum over 'w)
     */
   def over(alias: Expression): Expression = {
-    expr match {
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/expressionDsl.scala]: 把 new UnresolvedOverCall移到case里面？保留原来的exception?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/package.scala [hunk 16]
```diff
@@ -21,8 +21,8 @@ import org.apache.flink.types.Row
 import org.apache.flink.api.scala._
 import org.apache.flink.api.scala.DataSet
 import org.apache.flink.streaming.api.scala.DataStream
-import org.apache.flink.table.api.scala.{StreamTableEnvironment => ScalaStreamTableEnv}
-import org.apache.flink.table.api.scala.{BatchTableEnvironment => ScalaBatchTableEnv}
+import org.apache.flink.table.api.scala.{BatchTableEnvironment => ScalaBatchTableEnv, StreamTableEnvironment => ScalaStreamTableEnv}
+import org.apache.flink.table.functions.TableFunction
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/package.scala]: unused import,整个文件的改动都可以revert了

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala [hunk 17]
```diff
@@ -27,6 +27,7 @@ import org.apache.flink.table.utils.{StreamTableTestUtil, TableTestBase}
 import org.junit.Test
 
 class GroupWindowTest extends TableTestBase {
+
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala]: revert

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/UDAGGExpression.scala (hunk 13 above)
Comment (dianfu) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/UDAGGExpression.scala]: +1
